src/extension.ts

```diff
@@ -3,52 +3,77 @@ import * as vscode from 'vscode';
 import * as path from "path";
 import * as fs from "fs";
 import * as globToRegExp from "glob-to-regexp";
+import simpleGit, {SimpleGit} from 'simple-git';
 
 const GIT_IDENTIFIER = ".git/HEAD";
 const BRANCH_PREFIX = "ref: refs/heads/";
 let protectedBranches: string[] = [];
 let suppressPopup = false;
 let lastBranch = "";
+let statusWarningText = "";
+let popupWarningText = "";
+let warningTooltip = "";
+let warningColor = "";
+let warnIfRegex = "";
+let warnIfMsg = "";
+let warnIfMsgColor = "";
 
-export function activate(context: vscode.ExtensionContext) {
-    const status = vscode.window.createStatusBarItem(vscode.StatusBarAlignment.Left, 9999);
-    status.text = "WARNING";
+let gitRootDir;
 
+export function activate(context: vscode.ExtensionContext) {
+    // Construct the status display
+    const statusLocal = vscode.window.createStatusBarItem(vscode.StatusBarAlignment.Left, 9999);
+    const statusRemote = vscode.window.createStatusBarItem(vscode.StatusBarAlignment.Left, 9999);
     console.log("Extension branch-warnings initializing");
 
-    updateConfigs(status);
-
-    const gitpath = path.join(locateGitPath(vscode.workspace.rootPath), ".git");
-    const headpath = path.join(gitpath, "HEAD");
+    refresh(statusLocal, statusRemote);
     
-        // Install the debugger
+    // Install the debugger
+    const gitPath = path.join(gitRootDir, ".git");
     let disposable = vscode.commands.registerCommand('gitbranchwarn.debug', () => {
 		// Display a message box to the user
-        vscode.window.showInformationMessage('Found a .git folder at: ' + gitpath);
+        vscode.window.showInformationMessage('Found a .git folder at: ' + gitPath);
 	});
 	context.subscriptions.push(disposable);
 
-    
-    const pattern = new vscode.RelativePattern(gitpath, "HEAD");
+    // Register to watch for config or file system changes and update alerts.
+    const pattern = new vscode.RelativePattern(gitPath, "HEAD");
     const watcher = vscode.workspace.createFileSystemWatcher(pattern);
     watcher.onDidChange(e => {
         console.log(".git/HEAD change detected");
-        updateBranch(status, e.fsPath);
+        refresh(statusLocal, statusRemote);
+    });
+    vscode.workspace.onDidChangeWorkspaceFolders(e => {
+        console.log("Workspace folder change detected");
+        refresh(statusLocal, statusRemote);
     });
     vscode.workspace.onDidChangeConfiguration(e => {
         console.log("Configuration change detected");
-        updateConfigs(status);
-        updateBranch(status, headpath);
-    });
-    updateBranch(status, headpath);
+        refresh(statusLocal, statusRemote);
+    }); 
+}
+
+function refresh(statusLocal:vscode.StatusBarItem, statusRemote:vscode.StatusBarItem) {
+    updateConfigs();
+    gitRootDir = locateGitPath(vscode.workspace.rootPath);
+    checkLocalBranch(statusLocal);
+    checkRemoteBranches(statusRemote)
```

Comment:
Checking remote branches on every change might become be a bit too frequent for large repositories. 
Maybe using a timer that is invoked every 15 minutes, and only checks if there has been a change in the settings/workspace/head, could be enough.

Reply:
It doesn't run on every change it only runs when: (1) the plugin is activated (2) when there is a configuration change (3) when you switch workspace folders and (4) if there is a file change to the file .git/HEAD.